### PR TITLE
(FACT-971) Implement memory resolver for AIX

### PR DESCRIPTION
**Pull request by branan**
For memory, we just use the `vminfo` system call. For swap, we query the ODM for all devices marked as "paging", and use the swapqry system call to get information about them.

This has the same ODM implementation as #1121 (in fact the same commit). I'll be sure to rebase them as appropriate to cover any review of the ODM classes.

### CODE REVIEW

File: lib/src/facts/aix/memory_resolver.cc [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+        vminfo info;
+        auto res = vmgetinfo(&info, VMINFO, sizeof(info));
+        if (res < 0) {
+            throw system_error(errno, system_category());
+        }
+        result.mem_total = info.memsizepgs * 4096;
+        result.mem_free = info.numfrb * 4096;
+
+        auto cu_at_query = odm_class<CuAt>::open("CuAt").query("value=paging and attribute=type");
```
Comment (MikaelSmith) [inline lib/src/facts/aix/memory_resolver.cc]: So `CuAt` comes from `cfgodm.h`, and has a data member named `name`?

File: lib/src/facts/aix/memory_resolver.cc [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    memory_resolver::data memory_resolver::collect_data(collection& facts)
+    {
+        data result;
+
+        vminfo info;
+        auto res = vmgetinfo(&info, VMINFO, sizeof(info));
+        if (res < 0) {
+            throw system_error(errno, system_category());
+        }
+        result.mem_total = info.memsizepgs * 4096;
```
Comment (MikaelSmith) [inline lib/src/facts/aix/memory_resolver.cc]: Are these 4KiB sizes fixed, or could they depend on OS configuration?

Comment (branan) [inline lib/src/facts/aix/memory_resolver.cc]: page size is fixed at 4K. no good doc link - it's in comments in the various headers :(

File: lib/src/facts/aix/memory_resolver.cc (hunk 1 above)
Comment (branan) [inline lib/src/facts/aix/memory_resolver.cc]: Yeah. the odm_class type just wraps an existing ODM struct. cfgodm provides the ODM structs used in the AIX system configuration, including CuAt.

File: lib/src/facts/aix/memory_resolver.cc (hunk 2 above)
Comment (branan) [inline lib/src/facts/aix/memory_resolver.cc]: I should add that IBM's documentation points to the comments in the headers as being authoritative, so I trust it's not going to change anytime soon

**Comment by branan** [comment]
@MikaelSmith updated to use the PAGE_SIZE define from sys/limits.h

**Comment by MikaelSmith** [comment]
:+1:

File: lib/inc/internal/util/aix/odm.hpp [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+     * can open a single ODM class only once, and concurrent
+     * operations are not supported.
+     *
+     * @tparam T the struct that is stored in the ODM class.
+     */
+    template <typename T>
+    class odm_class : public std::enable_shared_from_this<odm_class<T>> {
+    public:
+        /// shared_ptr to an odm_class
+        typedef std::shared_ptr<odm_class<T>> ptr;
```
Comment (MikaelSmith) [inline lib/inc/internal/util/aix/odm.hpp]: Could also switch this to `using ptr = ...`.

File: lib/inc/internal/util/aix/odm.hpp [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+                     owner->locked = true;
+                 } else {
+                     // In theory nobody should be constructing us with
+                     // null data and valid owner, but why take the risk?
+                     owner = ptr(nullptr);
+                 }
+             }
+
+        private:
+             T *data;
```
Comment (MikaelSmith) [inline lib/inc/internal/util/aix/odm.hpp]: These should use `_` prefix/suffix.

File: lib/inc/internal/util/aix/odm.hpp [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+        protected:
+             /**
+              * Construct a query_proxy for an odm_class query
+              * @param q the query string
+              * @param o the odm_class that owns this query
+              */
+             query_proxy(std::string q, ptr o) : query(q), owner(o) {}
+
+        private:
+             std::string query;
```
Comment (MikaelSmith) [inline lib/inc/internal/util/aix/odm.hpp]: Should use `_`.

File: lib/inc/internal/util/aix/odm.hpp [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+
+        protected:
+             /**
+              * Construct a proxy for an odm_class
+              * @param self The odm_class that we proxy for
+              */
+             proxy(ptr self) : self(self) {}
+
+        private:
+             ptr self;
```
Comment (MikaelSmith) [inline lib/inc/internal/util/aix/odm.hpp]: Should use `_`.

File: lib/inc/internal/util/aix/odm.hpp [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+
+    protected:
+        /// The ODM class that we are wrapping
+        CLASS_SYMBOL klass;
+
+        /// whether we are currently locked for iteration
+        bool locked;
+
+    private:
+        odm::ptr the_odm;
```
Comment (MikaelSmith) [inline lib/inc/internal/util/aix/odm.hpp]: `_`? Also not sure on protected members, probably also `_`.

**Comment by branan** [comment]
@MikaelSmith underscores everywhere.

Alsoadjusted to enable perfect forwarding of the string argument in query creation. We were relying on inlining to avoid several string copies there. I'm pretty sure that'd all get inlined in an optimized build, but it was easy enough to explicitly use `std::forward`

File: lib/inc/internal/util/aix/odm.hpp [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+        }
+
+    protected:
+        /**
+         * @see proxy::query
+         * @param query the query string
+         * @return a query_proxy representing the provided query
+         */
+        template <typename Arg>
+       query_proxy query(Arg&& query) {
```
Comment (MikaelSmith) [inline lib/inc/internal/util/aix/odm.hpp]: Indentation.

**Comment by branan** [comment]
Indentation fixed

**Pull request merged**